typing/typedecl.ml

```diff
@@ -2353,6 +2367,41 @@ let add_types_to_env decls shapes env =
       add_type ~check:true ~shape id decl env)
     decls shapes env
 
+let mark_decl_jkind_best decl =
+  match decl.type_kind with
+  | Type_abstract _ | Type_open -> decl
+  | Type_record (_, _) | Type_record_unboxed_product (_, _) | Type_variant (_, _) ->
+    { decl with type_jkind = Jkind.mark_best decl.type_jkind }
+
+(* Normalize the jkinds in a list of (potentially mutually recursive) type declarations *)
+let normalize_decl_jkinds env shapes decls =
+  (* Mark the jkinds in each decl as best (as we have now added all the information we
+     need to to mark them best) *)
+  let decls = List.map (fun (id, decl) -> (id, mark_decl_jkind_best decl)) decls in
+  (* Add the types, with non-normalized kinds, to the environment to start, so that eg
+     types can look up their own (potentially non-normalized) kinds *)
+  let env = add_types_to_env decls shapes env in
+  let (decls, env) =
+    List.fold_left2
+      (fun (decls, env) (id, decl) shape ->
+         let normalized_jkind =
+           Jkind.normalize
+             ~require_best:true
+             ~jkind_of_type:(fun ty -> Some (Ctype.type_jkind env ty))
+             decl.type_jkind
+         in
+         let decl = { decl with type_jkind = normalized_jkind } in
+         (* Add the decl with the normalized kind back to the environment, so that later
+            kinds don't have to normalize this kind if they mention this type in their
+            with-bounds *)
+         let env = add_type ~check:false ~shape:shape id decl env in
```

Comment:
Similar to my comment above, this approach is not robust in the face of mutual recursion. That is, changing the order of declarations within a mutually recursive group will change the normalization behavior. Maybe that's OK? But at least it would be good to have a comment saying so. And if it affects compiler behavior (maybe requiring `module type of`?), there should be a test demonstrating the order-dependence.

Reply:
indeed. the actual correct thing to do is a little more involved, but seems to work fine.